### PR TITLE
Ultimate Access log filter

### PR DESCRIPTION
**Pull request by mmatur**
### What does this PR do?

This PR is an implementation of #2875.

### Motivation

Be able to customize accesslog like it was asked over PRs #2853 and #2397

Closes #2853, #2397, #2875

### More

- [x] Added/updated tests
- [x] Added/updated documentation

### CODE REVIEW

File: types/logs_test.go [hunk 1]
```diff
@@ -0,0 +1,403 @@
...
+)
+
+func TestStatusCodesSet(t *testing.T) {
+	tests := []struct {
+		desc     string
+		value    string
+		expected StatusCodes
+	}{
+		{
+			desc:     "One value should return StatusCodes of size 1",
```
Comment (marco-jantke) [inline types/logs_test.go]: I think the "should return StatusCodes ..." stuff can be dropped from the test descriptions.

File: types/logs_test.go [hunk 2]
```diff
@@ -0,0 +1,403 @@
...
+
+			var statusCodes StatusCodes
+			err := statusCodes.Set(test.value)
+			assert.Nil(t, err)
+			assert.Equal(t, test.expected, statusCodes)
+		})
+	}
+}
+
+func TestStatusCodesGet(t *testing.T) {
```
Comment (marco-jantke) [inline types/logs_test.go]: Do we really need to test the Getter? :)

File: types/logs_test.go (hunk 1 above)
Comment (ldez) [reply types/logs_test.go]: I don't agree with that.

**Comment by mmatur** [comment]
@marco-jantke Thanks for your review. I just push some fixes :)

File: types/logs_test.go (hunk 1 above)
Comment (marco-jantke) [reply types/logs_test.go]: I'm not super hot about this one, I just think it's good to decrease optical "noise" and redundant explanations.

File: types/logs_test.go (hunk 2 above)
Comment (marco-jantke) [reply types/logs_test.go]: Same here, this won't block my approval but adding ~30 lines of code to test one line that it is in fact super trivial and always does the same doesn't seem worth it.

File: types/logs.go [hunk 3]
```diff
@@ -0,0 +1,185 @@
...
+func (f *FieldsHeadersNames) SetValue(val interface{}) {
+	*f = val.(FieldsHeadersNames)
+}
+
+// Keep check if the field need to be kept or dropped
+func (f *Fields) Keep(field string) bool {
+	defaultKeep := true
+	if f != nil {
+		defaultKeep = checkFieldValue(f.DefaultMode, defaultKeep)
+		v, ok := f.Names[field]
```
[the remark this answers is not in the document]
Comment (marco-jantke) [reply types/logs.go]: An oversight here or do you disagree?

**Comment by marco-jantke** [comment]
@mmatur one thing I just missed/forgot again. What do you think about the comment I brought up now two times to drop redundant fields? See https://github.com/containous/traefik/pull/2988#pullrequestreview-103439929

**Comment by danieladams456** [comment]
@marco-jantke I am in favor. Seems like that would be follow-up PR, right?

**Comment by ldez** [comment]
@marco-jantke @danieladams456  we think this be breaking for people who parse the log output.
For now, we will not change fields.

I added the need in this issue https://github.com/containous/traefik/issues/2212

**Pull request merged**